software/led_color_cycle/led_color_cycle.c

```diff
@@ -0,0 +1,144 @@
+#include <stdio.h>
+#include "platform.h"
+
+// Do not set this to high, it may damage your led
+const uint8_t LED_MAX_BRIGHTNESS = 0x20;
+
+
+// Simple variables for LEDs, buttons, etc.
+volatile unsigned int* g_output_vals  = (unsigned int *) (GPIO_BASE_ADDR + GPIO_OUTPUT_VAL);
+volatile unsigned int* g_input_vals   = (unsigned int *) (GPIO_BASE_ADDR + GPIO_INPUT_VAL);
+volatile unsigned int* g_output_en    = (unsigned int *) (GPIO_BASE_ADDR + GPIO_OUTPUT_EN);
+volatile unsigned int* g_pullup_en    = (unsigned int *) (GPIO_BASE_ADDR + GPIO_PULLUP_EN);
+volatile unsigned int* g_input_en     = (unsigned int *) (GPIO_BASE_ADDR + GPIO_INPUT_EN);
+volatile unsigned int* g_iof_en       = (unsigned int *) (GPIO_BASE_ADDR + GPIO_IOF_EN);
+volatile unsigned int* g_iof_sel      = (unsigned int *) (GPIO_BASE_ADDR + GPIO_IOF_SEL);
+volatile unsigned int* g_out_xor      = (unsigned int *) (GPIO_BASE_ADDR + GPIO_OUTPUT_XOR);
+
+
+/**
+ * Wait a bit, TODO improve
+ */
+void busy_loop(uint32_t const duration) {
```

Comment:
I like to use the RTC, by reading the mtime register in the CLINT to do this sort of delay. Or for more precise loops, you can use mcycle (but then you need to know your CPU freq).